go/runtime/txpool/txpool.go

```diff
@@ -460,20 +466,49 @@ func (t *txPool) checkTxBatch(ctx context.Context, rr host.RichRuntime) {
 		return
 	}
 
+	// Check if node is synced.
+	var notReady bool
+	select {
+	case <-t.consensus.Synced():
+		// Node is synced.
+	default:
+		// Node is not synced.
+		notReady = true
+		t.logger.Debug("consensus is not synced yet")
+	}
+
+	if !notReady {
+		// Node is synced, ensure block round is synced to storage.
+		waitSyncCtx, cancelWaitSyncCtx := context.WithTimeout(ctx, checkTxWaitRoundSyncedTimeout)
+		defer cancelWaitSyncCtx()
+
+		t.logger.Debug("ensuring block round is synced", "round", bi.RuntimeBlock.Header.Round)
+		if _, err = t.history.WaitRoundSynced(waitSyncCtx, bi.RuntimeBlock.Header.Round); err != nil {
```

Comment:
If the block is synced and we call this method, do we need to wait for 6 seconds for the next block to be committed so that the `blocksNotifier` can broadcast the last block? I think that the history doesn't respond immediately as it doesn't track the latest synced round.

Why should txs check wait for round sync?

I would reorganize the code so that this call is not indented in `if`, because it looks like the happy path should  always check this. And I would think of blocking here until the round is synced. I think that currently it is possible for the check to never execute, if syncing takes too long. The reason is that the check call will fail because the block is not synced yet, but the next call will wait for a higher block to be synced, so it will need to wait and fail again.

Reply:
So maybe this function should start like this:
```go
bi, lastSyncedBlockProcessed, err := t.getLastSyncedBlockInfo()
```

Reply:
Hmm.  I will think about this tomorrow.

Reply:
> I think that the history doesn't respond immediately as it doesn't track the latest synced round.

In case the node has local storage available, history also waits for storage sync.

> Why should txs check wait for round sync?

Because otherwise it can't do the checks since the state will not be available and accessing the trie will fail.

Reply:
Did you think about the comment above (that we always need to wait up to 6s)?

Reply:
So I would update the `runtimeHistory` to store the last synced block into a local variable (probably locked) in `Commit`, and then in `WaitRoundSynced` first check the local variable, and then listen to blocks.

Reply:
As we discussed in the meeting, there is no 6s delay when the node is synced, since the history pubsub broker is instantiated with `publishLastOnSubscribe` set to `true` (https://github.com/oasisprotocol/oasis-core/blob/master/go/runtime/history/history.go#L411), so it will get the last block when subscribing :)

Reply:
The `blockNotifier` pushes the last block immediately, so my comment is way off and can be ignored. Resolving.